### PR TITLE
Move CompCert into coqPackages.

### PR DESCRIPTION
**Pull request by Zimmi48**
###### Motivation for this change

Move CompCert into `coqPackages` to make it easy to override using the standard `mkCoqDerivation` features.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/coq-modules/compcert/default.nix [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  releaseRev = v: "v${v}";
+
+  defaultVersion =  with versions; switch coq.version [
+      { case = range "8.8" "8.11";  out = "3.7"; }
+      { case = range "8.12" "8.13"; out = "3.8"; }
+    ] null;
+
+  release = {
+    "3.7".sha256 = "1h4zhk9rrqki193nxs9vjvya7nl9yxjcf07hfqb6g77riy1vd2jr";
+    "3.8".sha256 = "1gzlyxvw64ca12qql3wnq3bidcx9ygsklv9grjma3ib4hvg7vnr7";
```
Comment (vbgl) [inline pkgs/development/coq-modules/compcert/default.nix:24]: Where is `3.9` gone?

Comment (Zimmi48) [reply pkgs/development/coq-modules/compcert/default.nix:24]: Oops.

Comment (Zimmi48) [reply pkgs/development/coq-modules/compcert/default.nix:24]: Fixed and now the default unqualified attribute `compcert` is set at version 3.9 as before.

**Comment by Zimmi48** [comment]
Testing in coq-community/coq-nix-toolbox#35 has shown that CompCert 3.7 fails to build. It errors with:

```

OCAMLC   cparser/pre_parser.mli
OCAMLC   cparser/Cflow.mli
File "cparser/pre_parser.mli", line 114, characters 10-56:
Error: Unbound module MenhirLib
make[2]: *** [Makefile.extr:117: cparser/pre_parser.cmi] Error 2
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/build/source'
make[1]: *** [Makefile:162: ccomp] Error 2
make[1]: Leaving directory '/build/source'
make: *** [Makefile:138: all] Error 2
builder for '/nix/store/ws3iklq0zgazba0jrf6jw8gmx7ljfik2-coq8.10-compcert-3.7.drv' failed with exit code 2
error: build of '/nix/store/ws3iklq0zgazba0jrf6jw8gmx7ljfik2-coq8.10-compcert-3.7.drv' failed
Error: Process completed with exit code 100.
```

**Comment by Zimmi48** [comment]
So it turns out that this error was already present in `master`... I just ran `NIXPKGS_ALLOW_UNFREE=1 nix build -f . coqPackages_8_11.VST` and got:

```
builder for '/nix/store/jp4xk0pnkkb57vbz50g1p025aygi2h8j-compcert-3.7.drv' failed with exit code 2; last 10 log lines:
  File "cparser/pre_parser.mli", line 114, characters 10-56:
  114 |   include MenhirLib.IncrementalEngine.INCREMENTAL_ENGINE
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Error: Unbound module MenhirLib
  make[2]: *** [Makefile.extr:117: cparser/pre_parser.cmi] Error 2
```

**Comment by Zimmi48** [comment]
CompCert 3.8+ includes a new `-use-external-MenhirLib` configure option which we do not use in the derivation. This might be what makes the compilation pass for these versions.

**Comment by Zimmi48** [comment]
Indeed, I've confirmed this hypothesis with this patch:

```diff
diff --git a/pkgs/development/coq-modules/compcert/default.nix b/pkgs/development/coq-modules/compcert/default.nix
index fe7a6b325ed..69de417fae3 100644
--- a/pkgs/development/coq-modules/compcert/default.nix
+++ b/pkgs/development/coq-modules/compcert/default.nix
@@ -64,7 +64,7 @@ let compcert = mkCoqDerivation rec {
   };
 }; in
 compcert.overrideAttrs (o:
-  let useExternalFlocq = with lib.versions; range "3.8" "3.9" o.version; in
+  let useExternal = with lib.versions; range "3.8" "3.9" o.version; in
   {
     patches = with versions; switch [ coq.version o.version ] [
       { cases = [ (range "8.11" "8.11") "3.7" ];
@@ -110,14 +110,14 @@ compcert.overrideAttrs (o:
       }
     ] [];
 
-    propagatedBuildInputs = optional useExternalFlocq [ flocq ];
+    propagatedBuildInputs = optional useExternal [ flocq ];
 
     configurePhase = ''
       ./configure -clightgen \
       -prefix $out \
       -coqdevdir $lib/lib/coq/${coq.coq-version}/user-contrib/compcert/ \
       -toolprefix ${tools}/bin/ \
-      ${optionalString useExternalFlocq "-use-external-Flocq"} \
+      ${optionalString useExternal "-use-external-Flocq -use-external-MenhirLib"} \
       ${if stdenv.isDarwin then "x86_64-macosx" else "x86_64-linux"}
     '';
   }
```

Running `NIXPKGS_ALLOW_UNFREE=1 nix build -f . coqPackages_8_12.compcert` leads to a failure:

```
builder for '/nix/store/1bilyyk0inl54wrq37shv783v7v2rm4l-coq8.12-compcert-3.8.drv' failed with exit code 2; last 10 log lines:
  COQC lib/Wfsimpl.v
  COQC cparser/Parser.v
  File "./cparser/Parser.v", line 11, characters 23-27:
  Error: Cannot find a physical path bound to logical path matching suffix
  <> and prefix MenhirLib.
  
  make[1]: *** [Makefile:224: cparser/Parser.vo] Error 1
  make[1]: *** Waiting for unfinished jobs....
  make[1]: Leaving directory '/build/source'
  make: *** [Makefile:155: all] Error 2
[0 built (1 failed), 2 copied (471.0 MiB), 82.1 MiB DL]
error: build of '/nix/store/1bilyyk0inl54wrq37shv783v7v2rm4l-coq8.12-compcert-3.8.drv' failed
```

In passing, building with 8.13 leads to a different failure:

```
  make[1]: *** [Makefile:224: cparser/Parser.vo] Error 1
  make[1]: *** Waiting for unfinished jobs....
  File "./lib/Coqlib.v", line 1041, characters 0-76:
  Warning: The default value for hint locality is currently "local" in a
  section and "global" otherwise, but is scheduled to change in a future
  release. For the time being, adding hints outside of sections without
  specifying an explicit locality is therefore deprecated. It is recommended to
  use "export" whenever possible. [deprecated-hint-without-locality,deprecated]
  make[1]: Leaving directory '/build/source'
  make: *** [Makefile:155: all] Error 2
```

This looks like yet another bug (warnings should not be fatal and there is no apparent reason why this change would make them so). cc @xavierleroy

**Comment by xavierleroy** [comment]
Menhir >= 20210419 breaks compatibility with CompCert <= 3.8.   Use CompCert 3.9 or an older Menhir.

Your log show a failure on `cparser/Parser.vo` that has nothing to do with the warnings on `/lib/Coqlib.v`.

**Comment by Zimmi48** [comment]
Right. Thanks for the info and sorry for reading my build log too fast. @vbgl: Can we reintroduce a version of Menhir below `20210419`?

**Comment by Zimmi48** [comment]
An alternative is just to remove the currently broken versions (CompCert 3.7 and VST 2.6). This is what the last commit in this PR does. This means that VST is no longer available for Coq < 8.12. Currently testing at https://github.com/coq-community/coq-nix-toolbox/pull/35.

**Comment by vbgl** [comment]
Yes. Removing CompCert 3.7 seems to be the way to go!

**Comment by CohenCyril** [comment]
Well if you all agree, it's ok then... I guess if someone really wants VST on Coq < 8.13 they could PR a way to do it without breaking everything.

**Review by CohenCyril — APPROVED** [review]

**Comment by Zimmi48** [comment]
For the record, I tested the build of CompCert 3.9 locally and the build of CompCert 3.7 + VST was tested on the coq-nix-toolbox CI. So I think this PR is ready to merge.

**Comment by vbgl** [comment]
Thanks!

**Comment by Zimmi48** [comment]
My point for doing the `git mv` in a separate commit was to preserve the git history (not having `git blame` show as if I had written the whole file). But never mind: I should have explicitly requested a non-squash merge.

**Comment by vbgl** [comment]
Oops, sorry about that. I’ll try to ask before squashing in the future.

**Comment by Zimmi48** [comment]
This really is my fault. I should have known that this is the usual merging method you use here.

**Pull request merged**